man/tcos-volume-manager.1
.TH tcos-volume-manager 1 "Mar 11, 2007" "tcos-volume-manager man page"

.SH NAME
tcos-volume-manager \- Python GTK2 gui that allow to change thin client sound levels.

.SH DESCRIPTION

.PP
\fBtcos-volume-manager\fP is a python GUI exec in client session startup and allow to:
 * Read thin client sound card levels.
 * Change (mute/unmute) levels.

tcos-volume-manager use xauth with tcosxmlrpc (MIT-MAGIC-COOKIE)

.SH OPTIONS
.B [no options]
 With no options will launch hidden GUI (systray) app.

.B --host=xx.xx.xx.xx
 Overwrite DISPLAY with this var (not work unless thin client Xorg is started with \-ac)

.B --debug
 Show verbose output when running

.B --help
 Some usage help about tcos\-volume\-manager

.SH TIPS
 You can create 2 files to change default behaviour of tcos-volume-manager

.B ~/.tcos-volume-manager.allow.restartpulse
 This file show restart PulseAudio button

.B ~/.tcos-volume-manager.start.restartpulse
 This file restart PulseAudio when tcos-volume-manager starts


.SH BUGS
Contact Mario Izquierdo <mariodebian@gmail.com>

Homepage: http://www.tcosproject.org

.SH SEE ALSO
tcos-volume-manager(1)

.SH AUTHOR
Mario Izquierdo (mariodebian) <mariodebian@gmail.com>


Changelog
tcosmonitor (0.1.18.1) unstable; urgency=low

  * Move dbus conf file and tcos-dbus-client to tcos-tcosmonitor

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 07 Jul 2007 14:02:41 +0200

tcosmonitor (0.1.18.0) unstable; urgency=low

  * Remove pyexamples dir
  * xmlrpc/sh/restartx.sh:
    - call restartxorg instead of killing X
  * tcos-devices-ng:
    - Allot to disable floppy cdroms or hdd from conf file
      (disable_floppy=1, disable_cdroms=1, disable_hdd=1)
  * xmlrpc/*:
    - Add tcos.who method to retrieve username connected at :0

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 07 Jul 2007 13:28:15 +0200

tcosmonitor (0.1.17.2) unstable; urgency=low

  * xmlrpc/Makefile xmlrpc/var/etc/abyss.conf:
    - Set correct paths for tcos-standalone and tcos thin client

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 27 Jun 2007 23:54:22 +0200

tcosmonitor (0.1.17.1) unstable; urgency=low

  * New package tcos-standalone (Use tcosmonitor with not standalone hosts)
  * xmlrpc/*.h:
    - Add TCOS_PATH define to tcosxmlrpc, to allow running external tools from other paths
  * xmlrpc/sh/*:
    - Ad conditions to exec or not some tools.
  * TcosTrayIcon.py:
    - Remove StatusIcon declaration (break at gtk < 2.10)
  * Add support of standalone mode in TcoaActions.py
  * TcosActions.py:
    - patch to start/stop video broadcast (vidal_joshur at gva dot es)
    - patch to send files with rsync (vidal_joshur at gva dot es)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 27 Jun 2007 20:10:18 +0200

tcosmonitor (0.1.16.2) unstable; urgency=low

  * udev/tcos-udevd.sh:
    - Force udev to save ID_FS_TYPE in disk, not only partitions

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 20 Jun 2007 21:18:12 +0200

tcosmonitor (0.1.16.1) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Add support to mount NTFS with ntfs-3g driver
  * tcos-devices-ng.py:
    - Mount ntfs-3g in a thread (fuse mounts exec a process)
  * TcosXmlRpc.py:
    - Don't fail in GetDevicesInfo (use try--except)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 20 Jun 2007 15:14:16 +0200

tcosmonitor (0.1.16) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Fix typo in include header
  * images/*cdrom*
    - new image based on gnome theme
  * tcos-devices-ng.py:
    - Fix translation strings (vars go outside _('%s')%var )
    - Add umount_all function called from exit
    - Fix USB device remount
    - Allow to specify mounts points with a new conf file ~/.tcos-devices-ng.conf
  * Add new man page: man/tcos-devices-ng.1

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 19 Jun 2007 18:34:18 +0200

tcosmonitor (0.1.15.5) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Option gethdd now returns all type disk except swap or iso9660

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 19 Jun 2007 12:40:51 +0200

tcosmonitor (0.1.15.4) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Add --gethdd param and return ext3 or vfat devices in /etc/fstab
  * tcos-devices-ng.py:
    - Read hdd device partitions in thin client and put in menu
  * images/hdd_*mount.png:
    - Add hdd icons
  * po/*:
    - Update translations (new hdd strings)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 18 Jun 2007 18:52:00 +0200

tcosmonitor (0.1.15.3) unstable; urgency=low

  * lockscreen/* :
    - Change copyright header
    - Compile with imlib-config flags and libs
  * dbus/81tcos-utils:
    - Start tcos-devices --daemon and tcos-devices --hidden
  * tcos-devices-ng.py:
    - Don't use shared vars before importing module shared

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 18 Jun 2007 11:06:10 +0200

tcosmonitor (0.1.15.2) unstable; urgency=low

  * conf/tcos.conf man/* :
    - fix some english typos (close ticket #3) thanks to Andres Jimenez again
  * lockscreen/*:
    - Add support to put a png/jpg, xpm, etc... image in lockscreen
    - Default image locked.png
  * tcos-devices-ng.py:
    - Clean code and remote extra comments
  * pyexamples:
    - Add some python examples

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 17 Jun 2007 17:01:00 +0200

tcosmonitor (0.1.15.1) unstable; urgency=low

  * shared.py:
    - Remove some ugly debug info
  * xmlrpc/sh/devicesctl.sh:
    - Add --exists parameter to know if a file (or device) exists
  * xmlrpc/sh/soundctl.sh:
    - Exis if not soundcard found
  * xmlrpc/xauth.c:
    - Compare local hostname with cookie hostname.
  * tcos-devices.py:
    - Exit if X cookie don't match
  * images/*:
    - Add some tcos-devices-ng.py images
  * TcosTrayIcon.py:
    - New Tcos class (used to build systray menu)
  * tcos-devices-ng.py:
    - New and more powerfull remote device manager
  * Rename dbus/81tcos-dbus-client to dbus/81tcos-utils and start tcos-devices from there
  * po/*:
    - Update spanish translation

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Thu, 14 Jun 2007 21:15:15 +0200

tcosmonitor (0.1.14.4) unstable; urgency=low

  * TcosXauth.py:
    - Read DISPLAY info and search for hostname and IP address

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat,  9 Jun 2007 19:33:04 +0200

tcosmonitor (0.1.14.3) unstable; urgency=low

  * tcos-devices.py, tcos-volume-manager.py:
    - If gtk version is greater than 2.10 use gtk.statusicon instead of egg.trayicon

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 09 Jun 2007 17:11:33 +0200

tcosmonitor (0.1.14.2) unstable; urgency=low

  * TcosXauth.py:
    - Read xauth cookie with -n param to not resolve hostname
    - Try (with try: except: to get hostname with socket) not xauth locks

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Fri, 08 Jun 2007 22:15:23 +0200

tcosmonitor (0.1.14.1) unstable; urgency=low

  * Initialize.py:
    - Fix AboutDialog, close and destry button for gtk > 2.10

tcos-volume-manager.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
#
#    TcosMonitor version __VERSION__
#
# Copyright (c) 2006-2011 Mario Izquierdo <mariodebian@gmail.com>
#
# This package is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
#
# This package is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


import os, sys
import gobject
import getopt
from gettext import gettext as _

import pygtk
pygtk.require('2.0')
import gtk
from tcosmonitor import shared

# load conf file and exit if not active
if not shared.test_start("tcos-volume-manager") :
    print >> sys.stderr, "tcos-volume-manager disabled at %s" % (shared.module_conf_file)
    #print("tcos-volume-manager disabled at %s" % (shared.module_conf_file), file=sys.stderr)
    sys.exit(1)


import pwd

def print_debug(txt):
    if shared.debug:
        print >> sys.stderr, "%s::%s" % ("tcos-volume-manager", txt)
        #print("%s::%s" % ("tcos-volume-manager", txt), file=sys.stderr)

def get_username():
    return pwd.getpwuid(os.getuid())[0]


def usage():
    print ("tcos-volume-manager help:")
    print ("")
    print ("   tcos-volume-manager [--host=XXX.XXX.XXX.XXX] ")
    print ("                 (force host to connect to change volumes, default is DISPLAY)")
    print ("   tcos-volume-manager -d [--debug]  (write debug data to stdout)")
    print ("   tcos-volume-manager -h [--help]   (this help)")


try:
    opts, args = getopt.getopt(sys.argv[1:], ":hd", ["help", "debug", "host="])
except getopt.error, msg:
    print (msg)
    print ("for command line options use tcosconfig --help")
    sys.exit(2)

shared.remotehost=str(shared.parseIPAddress(os.environ["DISPLAY"]))

# process options
for o, a in opts:
    if o in ("-d", "--debug"):
        print ("DEBUG ACTIVE")
        shared.debug = True
    if o == "--host":
        shared.remotehost = str(shared.parseIPAddress(a))
    if o in ("-h", "--help"):
        usage()
        sys.exit()

if shared.remotehost == "":
    print ("tcos-volume-manager: Not allowed to run in local DISPLAY")
    #shared.error_msg ( _("tcos-volume-manager isn't allowed to run in local DISPLAY\nForce with --host=xx.xx.xx.xx") )
    sys.exit(0)



class TcosVolumeManager:
    def __init__(self, host):
        self.host=host
        self.name="TcosVolumeManager"
        self.visible=False
        
        if hasattr(gtk, 'status_icon_new_from_file'):
            # use gtk.status_icon
            icon = gtk.status_icon_new_from_file(shared.IMG_DIR + "tcos-volume-32x32.png")
            icon.set_tooltip( _("Tcos Sound levels on:\n%s") %(self.host) )
            icon.connect("activate", self.on_tray_icon_press_event)
        else:
            import egg.trayicon
            icon = egg.trayicon.TrayIcon("TCOS_sound")
            eventbox = gtk.EventBox()
            icon.add(eventbox)
            image=gtk.Image()
            image.set_from_file (shared.IMG_DIR + "tcos-volume-32x32.png")
            eventbox.add(image)
            tips = gtk.Tooltips() # deprecated
            
            tips.set_tip(icon, ( _("Tcos Sound levels on:\n%s") %(self.host) )[0:79])
            tips.enable()
            icon.show_all()
            eventbox.connect("button_press_event", self.on_tray_icon_press_event)
        
        
        from tcosmonitor.ping import PingPort
        if PingPort(self.host, shared.xmlremote_port, 0.5).get_status() != "OPEN":
            shared.error_msg( _("ERROR: It appears that TcosXmlRpc is not running on %s.") %(self.host) )
            sys.exit(1)
            
        
        import tcosmonitor.TcosXauth
        self.xauth=tcosmonitor.TcosXauth.TcosXauth(self)
        self.xauth.init_standalone()
        
        # get all channels
        import tcosmonitor.TcosXmlRpc
        import tcosmonitor.TcosConf
        self.config=tcosmonitor.TcosConf.TcosConf(self, openfile=False)
        self.xmlrpc=tcosmonitor.TcosXmlRpc.TcosXmlRpc(self)
        
        nossl=True
        # make a test and exit if no cookie match
        if not self.xauth.test_auth(nossl):
            print ("tcos-volume-manager: ERROR: Xauth cookie don't match")
            #sys.exit(1)
        
        self.xmlrpc.newhost(self.host,nossl)
        if not self.xmlrpc.connected:
            shared.error_msg( _("Error connecting with TcosXmlRpc in %s.") %(self.host) )
            sys.exit(1)
        
        # check for enabled sound
        have_sound=self.xmlrpc.IsEnabled("TCOS_SOUND")
        if not have_sound:
            print ("tcos-volume-manager: TCOS_SOUND is disabled")
            sys.exit(0)

        self.allchannels=self.xmlrpc.GetSoundChannelsContents()
        print_debug ("__init__() %s" %( self.allchannels ) )
        
        self.ui = gtk.Builder()
        self.ui.set_translation_domain(shared.PACKAGE)
        
        # Widgets
        self.ui.add_from_file(shared.GLADE_DIR + 'tcos-volume-manager.ui')
        self.mainwindow = self.ui.get_object('mainwindow')
        
        # close windows signals
        #self.mainwindow.connect('destroy', self.quitapp )
        self.mainwindow.connect('delete-event', self.mainwindow_close )
        
        self.mainlabel=self.ui.get_object('mainlabel')
        
        self.scrolledwindow=self.ui.get_object('scrolledwindow')
        self.scrolledwindow2=self.ui.get_object('scrolledwindow2')
        
        self.quitbutton=self.ui.get_object('quitbutton')
        self.quitbutton.connect('clicked', self.quitapp )
        
        self.statusbar=self.ui.get_object('statusbar')
        self.refreshbutton=self.ui.get_object('refreshbutton')
        self.refreshbutton.connect('clicked', self.on_refresh_button )
        
        self.restartbutton=self.ui.get_object('restartbutton')
        self.restartbutton.connect('clicked', self.on_restart_button )
        
        # only show if we found this file in $HOME dir
        allowfile=os.path.expanduser("~/.tcos-volume-manager.allow.restartpulse")
        if os.path.isfile(allowfile):
            self.restartbutton.show()
        
        # restart on start
        restartfile=os.path.expanduser("~/.tcos-volume-manager.start.restartpulse")
        if os.path.isfile(restartfile):
            self.xmlrpc.RestartSoundDaemon()
        
        # FIXME try to not focus on quitbutton
        self.refreshbutton.grab_focus()
        
        self.get_channel_info()
        
    def on_refresh_button(self, widget):
        self.allchannels=self.xmlrpc.GetSoundChannelsContents()
        self.scrolledwindow.foreach( self.delete_child, self.scrolledwindow )
        self.scrolledwindow2.foreach( self.delete_child, self.scrolledwindow2 )
        self.get_channel_info()

    def on_restart_button(self, widget):
        self.xmlrpc.RestartSoundDaemon()
        self.on_refresh_button(None)

    def delete_child(self, widget, scrolled):
        scrolled.remove(widget)

    def get_channel_info(self):
        # retry cookie auth
        nossl=True
        self.xauth.test_auth(nossl)
        
        primary_channels=[]
        secondary_channels=[]
        if self.allchannels != None and len(self.allchannels) > 0:
            gobject.timeout_add( 50, self.write_into_statusbar, _("Loading channels info...") )
            for channel in self.allchannels:
                #if not channel in shared.sound_only_channels:
                if not channel['name'] in shared.sound_only_channels:
                    secondary_channels.append(channel)
                    continue
                primary_channels.append(channel)
            gobject.timeout_add( 500, self.populate_mixer, primary_channels, self.scrolledwindow)
            gobject.timeout_add( 1500, self.populate_mixer, secondary_channels, self.scrolledwindow2)
            #gobject.timeout_add( 4000, self.write_into_statusbar, _("Ready") )
            
        else:
            print_debug ( "ERROR: No auth" )
            gobject.timeout_add( 2000, self.write_into_statusbar, _("Error loading channels info (xauth error)") )
        
        self.mainwindow.set_icon_from_file(shared.IMG_DIR + 'tcos-icon-32x32.png')
        self.mainwindow.set_title( _("Tcos Volume Manager")  )
        self.mainlabel.set_markup( _("<span size='large'><b>Sound mixer of %s host</b></span>") %(self.host) )
        
        #self.mainwindow.show()
        
        #self.populate_mixer(primary_channels, self.scrolledwindow)
        #self.populate_mixer(secondary_channels, self.scrolledwindow2)
        
    
    def populate_mixer(self, all_channels, scrollwindow):
        box1 = gtk.HBox(False, 0)
        box1.set_border_width(0)
        box1.show()
        scrollwindow.add_with_viewport(box1)
        
        for channel in all_channels:
            #frame = gtk.Frame(channel)
            frame = gtk.Frame(channel['name'])
            #frame.show()
            box2 = gtk.VBox(True, 0)
            box2.set_border_width(0)
            
            ###########################################################
            #value=self.xmlrpc.GetSoundInfo(channel, mode="--getlevel")
            #value=value.replace('%','')
            value=channel['level']
            try:
                value=float(value)
            except Exception, err:
                print_debug("Exception, can't get float value, error=%s"%err)
                value=0.0
            ctype=channel['type']
            #ismute=self.xmlrpc.GetSoundInfo(channel, mode="--getmute")
            ismute=channel['mute']
            if ismute == "off":
                ismute = True
            else:
                ismute = False
            #print_debug ( "populate_mixer() channel=%s ismute=%s volume level=%s" %(channel, ismute, value) )
            print_debug ( "populate_mixer() channel=%s ismute=%s volume level=%s ctype=%s" %(channel['name'], ismute, value, ctype) )
            #############################################################
            adjustment = gtk.Adjustment(value=0,
                                         lower=0,
                                         upper=100,
                                         step_incr=1,
                                         page_incr=1)
            volume_slider = None
            volume_slider = gtk.VScale(adjustment)
            volume_slider.set_digits(0)
            volume_slider.set_inverted(True)
            
            volume_slider.set_size_request(30, 100)
            volume_slider.set_value_pos(gtk.POS_TOP)     
            volume_slider.set_value( value )
            #volume_slider.connect("button_release_event", self.slider_value_changed, adjustment, channel, self.host)
            #volume_slider.connect("scroll_event", self.slider_value_changed, adjustment, channel, self.host)
            volume_slider.connect("button_release_event", self.slider_value_changed, adjustment, channel['name'], self.host)
            volume_slider.connect("scroll_event", self.slider_value_changed, adjustment, channel['name'], self.host)
            if "volume" in ctype:
                volume_slider.show()
            else:
                volume_slider.hide()
            
            box2.pack_start(volume_slider, False, True, 0)
            
            
            
            volume_checkbox=gtk.CheckButton(label=_("Mute"), use_underline=True)
            volume_checkbox.set_active(ismute)
            #volume_checkbox.connect("toggled", self.checkbox_value_changed, channel, self.host)
            volume_checkbox.connect("toggled", self.checkbox_value_changed, channel['name'], self.host)
            if "switch" in ctype:
                volume_checkbox.show()
            else:
                volume_checkbox.hide()
            
            
            box2.pack_start(volume_checkbox, False, True, 0)
            box2.show()
            frame.add(box2)
            
            frame.show()
            box1.pack_start(frame, True, True, 0)
        
        # write in statusbar if populating primary controls
        if scrollwindow == self.scrolledwindow:
            self.write_into_statusbar( _("Main controls ready") )
        
        # write in statusbar if populating secondary controls
        if scrollwindow == self.scrolledwindow2:
            self.write_into_statusbar( _("All remote controls loaded.") )
        return False
        

    def slider_value_changed(self, widget, event, adj, channel, ip):
        value=widget.get_value()
        print_debug ( "slider_value_changed() ip=%s channel=%s value=%s" %(ip, channel, value) )
        
        self.write_into_statusbar( \
        _("Changing value of %(channel)s channel, to %(value)s%%..." )\
         %{"channel":channel, "value":value} )
        
        tmp=self.xmlrpc.SetSound(ip, channel, str(value)+"%")
        newvalue="%2d%%"%int(tmp['level'])
        
        self.write_into_statusbar( \
        _("Changed value of %(channel)s channel, to %(value)s" ) \
        %{"channel":channel, "value":newvalue} )
    
    def checkbox_value_changed(self, widget, channel, ip):
        value=widget.get_active()
        if not value:
            value="off"
            self.write_into_statusbar( _("Unmuting %s channel..."  ) %(channel) )
            tmp=self.xmlrpc.SetSound(ip, channel, value="", mode="--setunmute")
            newvalue=tmp['mute']
        else:
            value="on"
            self.write_into_statusbar( _("Muting %s channel..."  ) %(channel) )
            tmp=self.xmlrpc.SetSound(ip, channel, value="", mode="--setmute")
            newvalue=tmp['mute']
        self.write_into_statusbar( _("Status of %(channel)s channel, is \"%(newvalue)s\""  )\
         %{"channel":channel, "newvalue":newvalue} ) 

    def write_into_statusbar(self, msg, *args):
        context_id=self.statusbar.get_context_id("status")
        self.statusbar.pop(context_id)
        self.statusbar.push(context_id, msg)
        return False

    def on_tray_icon_press_event(self, *args):
        if self.visible:
            self.mainwindow.hide()
        else:
            self.mainwindow.show()
        self.visible = not self.visible
        return
    
    def mainwindow_close(self, widget, event):
        print_debug ( "mainwindow_close() closing mainwindow to systray" )
        self.visible=False
        self.mainwindow.hide()
        return True

    def quitapp(self, *args):
        print_debug ( _("Exiting") )
        self.mainloop.quit()
        

    def run (self):
        self.mainloop = gobject.MainLoop()
        try:
            self.mainloop.run()
        except KeyboardInterrupt: # Ctrl+C Event
            self.quitapp()

if __name__ == "__main__":
    app=TcosVolumeManager(shared.remotehost)
    app.run()
